### PR TITLE
[GEP-26] Cloudprovider secret can contain workload identity data

### PR DESCRIPTION
**Pull request by dimityrmirchev**
**How to categorize this PR?**

/area security
/kind enhancement

**What this PR does / why we need it**:
It is now possible to reference a workload identity through the shoot spec. This will result in the `cloudprovider` secret deployed in the seed to contain metadata for the referenced workload identity.

cc @vpnachev 
**Which issue(s) this PR fixes**:
Fixes #

Part of https://github.com/gardener/gardener/issues/9586

**Special notes for your reviewer**:

**Release note**:

```other operator
NONE
```

### CODE REVIEW

**Comment by dimityrmirchev** [comment]
`make gardener-up` seems to fail now with the sort order introduced in https://github.com/gardener/gardener/blob/bf98fc33037a052e2d395787514c9c6dbfe33d40/example/provider-local/garden/skaffold/kustomization.yaml. Will check..

File: docs/deployment/gardenlet_api_access.md [hunk 1]
```diff
@@ -68,6 +68,7 @@ Today, the following rules are implemented:
...
 | `ShootState`                | `get`, `create`, `update`, `patch`                              | `ShootState` -> `Shoot` -> `Seed`                                                                                                                                                    | Allow only `get`, `create`, `update`, `patch` requests for `ShootState`s belonging by `Shoot`s that are assigned to the `gardenlet`'s `Seed`.                                                                                                                                    |
+| `WorkloadIdentity`          | `get`                                                           | `WorkloadIdentity` -> `CredentialsBinding` -> `Shoot` -> `Seed`                                                                                                                      | Allow only `get` requests for `WorkloadIdentities` referenced by `CredentialsBinding`s referenced by `Shoot`s that are assigned to the `gardenlet`'s `Seed`.                                                                                                                     |
```
Comment (rfranzke) [inline docs/deployment/gardenlet_api_access.md:71]: Please augment https://github.com/gardener/gardener/blob/master/docs/deployment/content/gardenlet_api_access_graph.png accordingly

File: example/provider-local/garden/base/credentialsbinding.yaml [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+metadata:
+  name: local
+  namespace: garden-local
+spec:
+  audiences:
+  - "provider-local"
+  targetSystem:
+    type: local
+    providerConfig:
+      foo: bar
```
Comment (rfranzke) [inline example/provider-local/garden/base/credentialsbinding.yaml]: This can be removed, I guess?

File: example/provider-local/garden/base/kustomization.yaml [hunk 3]
```diff
@@ -1,10 +1,14 @@
 apiVersion: kustomize.config.k8s.io/v1beta1
 kind: Kustomization
 
+sortOptions:
+  order: fifo
```
Comment (rfranzke) [inline example/provider-local/garden/base/kustomization.yaml:5]: Why is this (suddenly) needed?

File: example/provider-local/garden/local/kustomization.yaml [hunk 4]
```diff
@@ -2,8 +2,8 @@ apiVersion: kustomize.config.k8s.io/v1beta1
 kind: Kustomization
 
 resources:
-- controllerregistration.yaml
 - controllerdeployment.yaml
+- controllerregistration.yaml
```
Comment (rfranzke) [inline example/provider-local/garden/local/kustomization.yaml:6]: Why was this changed?

File: pkg/controllermanager/controller/credentialsbinding/reconciler.go [hunk 5]
```diff
@@ -60,6 +61,28 @@ func (r *Reconciler) Reconcile(ctx context.Context, request reconcile.Request) (
...
+	wiGVK := securityv1alpha1.SchemeGroupVersion.WithKind("WorkloadIdentity")
+	if ref.GroupVersionKind() == wiGVK {
+		workloadIdentity := &securityv1alpha1.WorkloadIdentity{}
+		if err := r.Client.Get(ctx, client.ObjectKey{Namespace: ref.Namespace, Name: ref.Name}, workloadIdentity); err != nil {
+			return nil, err
+		}
+		return workloadIdentity, nil
+	}
+
+	return nil, fmt.Errorf("unsuported credentials reference: %s, %s", ref.Namespace+"/"+ref.Name, ref.GroupVersionKind().String())
```
Comment (rfranzke) [inline pkg/controllermanager/controller/credentialsbinding/reconciler.go]:
```suggestion
	var obj client.Object

	switch ref.GroupVersionKind() {
	case corev1.SchemeGroupVersion.WithKind("Secret"):
		obj = &corev1.Secret{}
	case securityv1alpha1.SchemeGroupVersion.WithKind("WorkloadIdentity"):
		obj = &securityv1alpha1.WorkloadIdentity{}
	default:
		return nil, fmt.Errorf("unsuported credentials reference: %s, %s", ref.Namespace+"/"+ref.Name, ref.GroupVersionKind().String())
	}
	
	return obj, r.Client.Get(ctx, client.ObjectKey{Namespace: ref.Namespace, Name: ref.Name}, obj)
```

File: pkg/gardenlet/operation/botanist/secrets.go [hunk 6]
```diff
@@ -459,24 +461,62 @@ func (b *Botanist) reconcileWildcardIngressCertificate(ctx context.Context) erro
...
-		secret.Annotations = map[string]string{
-			"checksum/data": checksum,
+	var (
+		data                  map[string][]byte
+		additionalLabels      map[string]string
+		additionalAnnotations map[string]string
+	)
+	switch credentials := b.Shoot.Credentials.(type) {
+	case *securityv1alpha1.WorkloadIdentity:
+		providerConfigBytes, err := json.Marshal(credentials.Spec.TargetSystem.ProviderConfig)
```
Comment (rfranzke) [inline pkg/gardenlet/operation/botanist/secrets.go]:
```suggestion
		providerConfigRaw, err := json.Marshal(credentials.Spec.TargetSystem.ProviderConfig)
```

File: pkg/gardenlet/operation/botanist/secrets_test.go [hunk 7]
```diff
@@ -109,6 +116,74 @@ var _ = Describe("Secrets", func() {
...
+				},
+				Data: map[string][]byte{"config": []byte(`{"raw":"raw"}`)},
+				Type: corev1.SecretTypeOpaque,
+			}))
+		})
+
+		It("should return error when shoot credentials are of unknown type", func() {
+			botanist.Shoot.Credentials = &corev1.Pod{}
+			err := botanist.DeployCloudProviderSecret(ctx)
+			Expect(err.Error()).To(Equal("unexpected type *v1.Pod, should be either Secret or WorkloadIdentity"))
```
Comment (rfranzke) [inline pkg/gardenlet/operation/botanist/secrets_test.go]:
```suggestion
			Expect(botanist.DeployCloudProviderSecret(ctx)).To(MatchError(Equal("unexpected type *v1.Pod, should be either Secret or WorkloadIdentity")))
```

File: pkg/gardenlet/operation/shoot/shoot.go [hunk 8]
```diff
@@ -129,13 +138,20 @@ func (b *Builder) WithShootSecretFrom(c client.Reader) *Builder {
...
-			// TODO(dimityrmirchev): This code should eventually handle
-			// the credentials binding referencing a workload identity
 			binding := &securityv1alpha1.CredentialsBinding{}
 			if err := c.Get(ctx, client.ObjectKey{Namespace: namespace, Name: bindingName}, binding); err != nil {
 				return nil, err
 			}
 			key = client.ObjectKey{Namespace: binding.CredentialsRef.Namespace, Name: binding.CredentialsRef.Name}
+
+			wiGVK := securityv1alpha1.SchemeGroupVersion.WithKind("WorkloadIdentity")
+			if wiGVK == binding.CredentialsRef.GroupVersionKind() {
```
Comment (rfranzke) [inline pkg/gardenlet/operation/shoot/shoot.go]:
```suggestion
			if binding.CredentialsRef.GroupVersionKind() == securityv1alpha1.SchemeGroupVersion.WithKind("WorkloadIdentity") {
```

File: pkg/utils/gardener/shoot_test.go [hunk 9]
```diff
@@ -1027,6 +1028,54 @@ var _ = Describe("Shoot", func() {
...
+									Primary: ptr.To(true),
+								},
+							},
+						},
+					},
+				}
+			)
+
+			_, err := ConstructExternalDomain(ctx, fakeClient, shoot, pod, nil)
+			Expect(err.Error()).To(Equal("unexpected shoot credentials type"))
```
Comment (rfranzke) [inline pkg/utils/gardener/shoot_test.go]: same here

File: pkg/utils/gardener/shoot_test.go [hunk 10]
```diff
@@ -1027,6 +1028,54 @@ var _ = Describe("Shoot", func() {
...
+									Primary: ptr.To(true),
+								},
+							},
+						},
+					},
+				}
+			)
+
+			_, err := ConstructExternalDomain(ctx, fakeClient, shoot, workloadIdentity, nil)
+			Expect(err.Error()).To(Equal("shoot credentials of type workload identity cannot be used as domain secret"))
```
Comment (rfranzke) [inline pkg/utils/gardener/shoot_test.go]: same here

File: example/provider-local/garden/base/kustomization.yaml (hunk 3 above)
Comment (dimityrmirchev) [reply example/provider-local/garden/base/kustomization.yaml:5]: Kustomize by default applies resources in some legacy order that uses alphabetical sorting using groups + kinds. This means that `WorkloadIdentity` is after `CredentialsBinding`. I want kustomize to honour the order from the [credentialsbinding file](https://github.com/gardener/gardener/blob/0f11fd7cd17542cb6f258cd87447c4b02ed50552/example/provider-local/garden/base/credentialsbinding.yaml). If the order is not honoured then validation will fail because the referenced `WorkloadIdentity` does not exist.

File: example/provider-local/garden/local/kustomization.yaml (hunk 4 above)
Comment (dimityrmirchev) [reply example/provider-local/garden/local/kustomization.yaml:6]: Same reason as https://github.com/gardener/gardener/pull/10239#discussion_r1703953903 but I want `controllerdeployment` and `controllerregistration` ordered.

Comment (rfranzke) [reply example/provider-local/garden/local/kustomization.yaml:6]: Fair enough, but why was this working before/is needed now (with this PR)?

Comment (dimityrmirchev) [reply example/provider-local/garden/local/kustomization.yaml:6]: The behaviour changed when I added the ordering to https://github.com/gardener/gardener/blob/0f11fd7cd17542cb6f258cd87447c4b02ed50552/example/provider-local/garden/skaffold/kustomization.yaml#L4-L5

I guess that these settings are inherited and propagated to referenced kustomization files.
https://github.com/gardener/gardener/blob/0f11fd7cd17542cb6f258cd87447c4b02ed50552/example/provider-local/garden/skaffold/kustomization.yaml#L7-L9

File: pkg/gardenlet/operation/botanist/secrets_test.go [hunk 11]
```diff
@@ -109,6 +116,73 @@ var _ = Describe("Secrets", func() {
 		botanist.Shoot.SetShootState(&gardencorev1beta1.ShootState{})
 	})
 
+	Describe("#DeployCloudProviderSecret", func() {
+		It("should create cloud provider secret contaning secret data", func() {
```
Comment (vpnachev) [inline pkg/gardenlet/operation/botanist/secrets_test.go]:
```suggestion
		It("should create cloud provider secret containing secret data", func() {
```

File: plugin/pkg/global/resourcereferencemanager/admission.go [hunk 12]
```diff
@@ -639,39 +666,64 @@ func (r *ReferenceManager) ensureBindingReferences(ctx context.Context, attribut
...
+		APIGroup:        credentialsAPIGroup,
+		APIVersion:      credentialsAPIVersion,
+		Resource:        credentialsResource,
+		Namespace:       credentialsNamespace,
+		Name:            credentialsName,
 		ResourceRequest: true,
 	}
 	if decision, _, _ := r.authorizer.Authorize(ctx, readAttributes); decision != authorizer.DecisionAllow {
-		return fmt.Errorf("%s cannot reference a secret you are not allowed to read", binding.GetObjectKind().GroupVersionKind().Kind)
+		return fmt.Errorf("%s cannot reference a %s you are not allowed to read", credentialsKind, binding.GetObjectKind().GroupVersionKind().Kind)
```
Comment (vpnachev) [inline plugin/pkg/global/resourcereferencemanager/admission.go]: Shouldn't it be ?
```suggestion
		return fmt.Errorf("%s cannot reference a %s you are not allowed to read", binding.GetObjectKind().GroupVersionKind().Kind, credentialsKind)
```

**Review by vpnachev — COMMENTED** [review]
Some minor nits from my side

File: plugin/pkg/global/resourcereferencemanager/admission.go (hunk 12 above)
Comment (dimityrmirchev) [reply plugin/pkg/global/resourcereferencemanager/admission.go]: You are right

**Comment by dimityrmirchev** [comment]
@rfranzke @vpnachev comments should be addressed. PTAL

**Review by rfranzke — COMMENTED** [review]
/lgtm

**Comment by dimityrmirchev** [comment]
Seems that the upgrade tests will flake because the previous version of `resourcereferencemanager` admission plugin was working only with `credentialsbinding` -> `secret` references. This explains the following error message from [test run ](https://prow.gardener.cloud/view/gs/gardener-prow/pr-logs/pull/gardener_gardener/10239/pull-gardener-e2e-kind-ha-single-zone-upgrade/1820814987451437056)

```
Error from server (Forbidden): credentialsbindings.security.gardener.cloud "local" is forbidden: secrets "local" not found
```

The same test can pass depending on which version of the admission plugin validates the `credentialsbinding` resource. See [this run](https://prow.gardener.cloud/view/gs/gardener-prow/pr-logs/pull/gardener_gardener/10239/pull-gardener-e2e-kind-ha-single-zone-upgrade/1820795288370024448).

This also means that https://github.com/gardener/gardener/pull/10239/commits/da322dc63c79394078b1c8f9b00a9ed3a5df9cba is not actually needed.

I am not sure how to proceed. The upgrade tests only fail because we want to deploy a `credentialsbindings` with the new gardener setup make recipes. This should be no issue during real upgrade of Gardener.

I am not sure how to proceed with this one. One solution would be to include a temporary secret in `garden-local` namespace called `local` so that the upgrade targets pass (and remove it later). Or something else 🤔 @rfranzke WDYT?

/hold 

Holding, so I can revert https://github.com/gardener/gardener/pull/10239/commits/da322dc63c79394078b1c8f9b00a9ed3a5df9cba once we decide how to proceed.

**Comment by vpnachev** [comment]
> One solution would be to include a temporary secret in garden-local namespace called local so that the upgrade targets pass (and remove it later). Or something else

Alternatively, is it possible to disable the test where workload identity is used instead of secret in the `*-upgrade` targets temporary? It can be re-enabled after the changes in this PR are released.

**Comment by dimityrmirchev** [comment]
> Alternatively, is it possible to disable the test where workload identity is used instead of secret

Actually there is no such test. Just the resources `credentialsbinding` + `workloadidentity` are created in the local env for easier testing setup, but tests are not switched yet. But yes, we can disable it, although I would prefer that the resources are created with `make gardener-up` and not separately.

I do not mind any of the approaches.

**Comment by dimityrmirchev** [comment]
OK, tests seem to pass now. It only needs a lgtm label 

/unhold
